_posts/2024-04-20-websockets2.md

```diff
@@ -0,0 +1,1861 @@
+  ...
+    val delMessages: Command[RoomId] =
+      sql"""
+              DELETE FROM messages 
+              WHERE room_id = $roomId
+          """.command
+}
+```
+The usage is similar to `deleteUserOrRoom()` with the help of the `execute` method:
+```scala
+...
+new PostgresProtocol[F] {
+  ...
+  override def deleteRoomMessages(roomId: RoomId): F[Unit] =
+    postgres.use { session =>
+      session.prepare(delMessages).flatMap { cmd =>
+        cmd.execute(roomId).void
+      }
+    }
+}
+```
+
+## 6. The Redis and Chat Protocols
+Before we dive into the Redis implementation, we'll need an overview of how the schema will be. 
+A Redis hash is a record type structured as a collection of field-value pairs, we'll need the following hashes for our application:
+
+The `users` hash will be a collection of UUID string fields mapped to user name values.
+The `rooms` hash will be a collection of UUID string fields mapped to room name values.
+The `userroomid` hash will be a collection of user UUID string fields mapped to and room UUID string values.
+
+A Redis set is a collection of unique string values, we'll need a Redis set for each `room` created in our application:
+The sets will be named in the format, `room:<room id>`, and will be a collection of user UUID strings.
+
+We'll be creating both the Redis Algebra and Chat Algebra in this section so that we can follow why each Redis function is needed.
+
+First, we'll start by creating the Redis Algebra for our application, in the `websockets` folder create a `RedisPotocol.scala` file with the following contents:
+
+```scala
+package rockthejvm.websockets
+import rockthejvm.websockets.domain.user.*
+import rockthejvm.websockets.domain.room.*
+
+trait RedisProtocol[F[_]] {
+  def createUser(user: User): F[Unit]
+  def createRoom(room: Room): F[Unit]
+  def getRoomFromName(roomname: String): F[Option[Room]]
+  def getUsersRoomId(user: User): F[Option[RoomId]]
+  def usernameExists(username: String): F[Boolean]
+  def listUserIds(roomid: RoomId): F[Set[UserId]]
+  def listRooms: F[List[String]]
+  def roomExists(roomid: RoomId): F[Boolean]
+  def mapUserToRoom(userid: UserId, roomid: RoomId): F[Unit]
+  def addUserToRoom(userid: UserId, roomid: RoomId): F[Unit]
+  def removeUserFromRoom(roomid: RoomId, userid: UserId): F[Unit]
+  def deleteUserRoomMapping(userid: UserId): F[Unit]
+  def deleteRoom(roomid: RoomId): F[Long]
+  def deleteUser(userid: UserId): F[Long]
+  def getSelectedUsers(
+      userid: UserId,
+      rest: List[UserId]
+  ): F[Option[List[User]]]
+  def chatState: F[String]
+}
+```
+Just like with Postgres, we'll be implementing each of these functions, and by the end of this section, we will have learned alot about Redis and the redis4cats package.
+
+We'll also need to create the companion object for `RedisProtocol` along with a `make()` function.
+
+```scala
+import dev.profunktor.redis4cats.RedisCommands
+ ...
+
+object RedisProtocol {
+  def make[F[_]](
```

Comment:
Same comment about the effectful smart constructor - you'll need to return an F[RedisProtocol[F]] or Resource[F, RedisProtocol[F]]

Reply:
Updated to `F[RedisProtocol[F]]`, also did the same for `F[InputMessage[F]]`